File: web/src/api/audit.ts
// 操作审计 API 封装（列表检索 + 元数据）
import { http } from "@/utils/request";

export type AuditAction = "create" | "update" | "delete" | "other";

export interface AuditLogItem {
  id: number;
  user_id: number | null;
  username: string | null;
  dept_id: number | null;
  action: string;
  module: string;
  method: string;
  path: string;
  query: string | null;
  status_code: number;
  ip: string | null;
  detail: string | null;
  created_at: string | null;
}

export interface AuditLogPage {
  total: number;
  items: AuditLogItem[];
  page: number;
  size: number;
}

export interface AuditListParams {
  page?: number;
  size?: number;
  module?: string;
  action?: string;
  username?: string;
  start?: string;
  end?: string;
}

export interface AuditMeta {
  modules: string[];
  actions: string[];
}

export function fetchAuditLogs(params: AuditListParams = {}): Promise<AuditLogPage> {
  return http<AuditLogPage>({
    url: "/v1/audit-logs",
    method: "GET",
    params,
  });
}

export function fetchAuditMeta(): Promise<AuditMeta> {
  return http<AuditMeta>({
    url: "/v1/audit-logs/meta",
    method: "GET",
  });
}

/** 导出审计日志为 CSV 文件 */
export function exportAuditLogs(params: Omit<AuditListParams, "page" | "size"> = {}): void {
  const query = new URLSearchParams();
  if (params.module) query.set("module", params.module);
  if (params.action) query.set("action", params.action);
  if (params.username) query.set("username", params.username);
  if (params.start) query.set("start", params.start);
  if (params.end) query.set("end", params.end);
  window.open(`/api/v1/audit-logs/export?${query.toString()}`, "_blank");
}

File: app/schema/audit.py
"""操作审计 Schema：对外输出与分页。"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_serializer

from app.core.clock import LOCAL_TZ


class AuditLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int | None = None
    username: str | None = None
    dept_id: int | None = None
    action: str
    module: str
    method: str
    path: str
    query: str | None = None
    status_code: int
    ip: str | None = None
    detail: str | None = None
    created_at: datetime | None = None

    @field_serializer("created_at")
    def _serialize_dt(self, v: datetime | None) -> str | None:
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()


class AuditLogPage(BaseModel):
    total: int
    items: list[AuditLogOut]
    page: int = 1
    size: int = 20
